File: app/scala/src/main/scala/com/example/accounting/domain/financial/BalanceSheet.scala
package com.example.accounting.domain.financial

import java.time.LocalDate

/**
 * 貸借対照表（Balance Sheet / B/S）
 *
 * @param asOfDate 基準日
 * @param assets 資産項目
 * @param liabilities 負債項目
 * @param equity 純資産項目
 * @param totalAssets 総資産
 * @param totalLiabilities 総負債
 * @param totalEquity 総純資産
 * @param totalLiabilitiesAndEquity 負債・純資産合計
 */
case class BalanceSheet(
    asOfDate: LocalDate,
    assets: List[BalanceSheetItem],
    liabilities: List[BalanceSheetItem],
    equity: List[BalanceSheetItem],
    totalAssets: BigDecimal,
    totalLiabilities: BigDecimal,
    totalEquity: BigDecimal,
    totalLiabilitiesAndEquity: BigDecimal,
):
  /**
   * 貸借平均の原則が成立しているかを検証
   * （資産 = 負債 + 純資産）
   */
  def isBalanced: Boolean =
    totalAssets == totalLiabilities + totalEquity

File: app/ruby/accounting_system/app/models/balance_sheet.rb
# frozen_string_literal: true

# 貸借対照表（Balance Sheet / B/S）
class BalanceSheet
  attr_accessor :as_of_date, :assets, :liabilities, :equity,
                :total_assets, :total_liabilities, :total_equity,
                :total_liabilities_and_equity

  def initialize(
    as_of_date:,
    assets: [],
    liabilities: [],
    equity: [],
    total_assets: 0,
    total_liabilities: 0,
    total_equity: 0,
    total_liabilities_and_equity: 0
  )
    @as_of_date = as_of_date
    @assets = assets
    @liabilities = liabilities
    @equity = equity
    @total_assets = total_assets
    @total_liabilities = total_liabilities
    @total_equity = total_equity
    @total_liabilities_and_equity = total_liabilities_and_equity
  end
end
